bootstrap/kubeadm/internal/controllers/kubeadmconfig_controller.go

```diff
@@ -248,13 +248,18 @@ func (r *KubeadmConfigReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		log.Info("Cluster infrastructure is not ready, waiting")
 		conditions.MarkFalse(config, bootstrapv1.DataSecretAvailableCondition, bootstrapv1.WaitingForClusterInfrastructureReason, clusterv1.ConditionSeverityInfo, "")
 		return ctrl.Result{}, nil
-	// Reconcile status for machines that already have a secret reference, but our status isn't up to date.
-	// This case solves the pivoting scenario (or a backup restore) which doesn't preserve the status subresource on objects.
-	case configOwner.DataSecretName() != nil && (!config.Status.Ready || config.Status.DataSecretName == nil):
-		config.Status.Ready = true
-		config.Status.DataSecretName = configOwner.DataSecretName()
-		conditions.MarkTrue(config, bootstrapv1.DataSecretAvailableCondition)
-		return ctrl.Result{}, nil
+		// For machines that already have a secret reference ensure config status is up-to-date and the bootstrapSecret is adopted.
+		// This case solves the pivoting scenario (or a backup restore) which doesn't preserve the status subresource or bootstrapSecret ownerReferences on objects.
+	case configOwner.DataSecretName() != nil:
```

Comment:
I'm pretty sure this breaks the case below.

We know always end up in this case when `configOwner.DataSecretName() != nil`

This probably means we will never get into `case config.Status.Ready`

Can we please add the new logic above this switch case to make this a lot safer?